File: src/Syntax/JavaPackage.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using AutoRest.Java.Model;
using AutoRest.Java.Template;

namespace AutoRest.Java
{
    public class JavaPackage : IEnumerable<JavaFile>
    {
        private JavaSettings settings;
        private List<JavaFile> javaFiles;
        private JavaFileFactory javaFileFactory;

        public JavaPackage()
        {
            this.settings = JavaSettings.Instance;
            this.javaFiles = new List<JavaFile>();
            this.javaFileFactory = new JavaFileFactory(settings);
        }

        public void AddManager(string package, string name, Manager model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.ManagerTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddServieClient(string package, string name, ServiceClient model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.ServiceClientTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddServiceClientInterface(string name, ServiceClient model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(settings.Package, name);
            Templates.ServiceClientInterfaceTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddMethodGroup(string package, string name, MethodGroupClient model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.MethodGroupTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddMethodGroupInterface(string name, MethodGroupClient model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(settings.Package, name);
            Templates.MethodGroupInterfaceTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddModel(string package, string name, ClientModel model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.ModelTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddException(string package, string name, ClientException model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.ExceptionTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddEnum(string package, string name, EnumType model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.EnumTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddPage(string package, string name, PageDetails model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.PageTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddClientResponse(string package, string name, ClientResponse model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.ResponseTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddXmlSequenceWrapper(string package, string name, XmlSequenceWrapper model)
        {
            var javaFile = javaFileFactory.CreateSourceFile(package, name);
            Templates.XmlSequenceWrapperTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddPackageInfo(string package, string name, PackageInfo model)
        {
            var javaFile = javaFileFactory.CreateEmptySourceFile(package, name);
            Templates.PackageInfoTemplate.Write(model, javaFile);
            javaFiles.Add(javaFile);
        }

        public void AddPom(PomTemplate pomTemplate)
        {
            StringBuilder pomContentsBuilder = new StringBuilder();
            using (pomTemplate.TextWriter = new StringWriter(pomContentsBuilder))
            {
                pomTemplate.ExecuteAsync().GetAwaiter().GetResult();
            }
            javaFiles.Add(new JavaFile("pom.xml", pomContentsBuilder.ToString()));
        }

        public IEnumerator<JavaFile> GetEnumerator()
        {
            return javaFiles.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return javaFiles.GetEnumerator();
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/model/javamodel/JavaPackage.java
package com.azure.autorest.model.javamodel;

import com.azure.autorest.extension.base.plugin.JavaSettings;
import com.azure.autorest.model.clientmodel.ClientException;
import com.azure.autorest.model.clientmodel.ClientModel;
import com.azure.autorest.model.clientmodel.ClientResponse;
import com.azure.autorest.model.clientmodel.EnumType;
import com.azure.autorest.model.clientmodel.Manager;
import com.azure.autorest.model.clientmodel.MethodGroupClient;
import com.azure.autorest.model.clientmodel.PackageInfo;
import com.azure.autorest.model.clientmodel.PageDetails;
import com.azure.autorest.model.clientmodel.ServiceClient;
import com.azure.autorest.model.clientmodel.XmlSequenceWrapper;
import com.azure.autorest.template.Templates;

import java.util.ArrayList;

public class JavaPackage
{
    private JavaSettings settings;
    private ArrayList<JavaFile> javaFiles;
    private JavaFileFactory javaFileFactory;

    public JavaPackage()
    {
        this.settings = JavaSettings.getInstance();
        this.javaFiles = new ArrayList<JavaFile>();
        this.javaFileFactory = new JavaFileFactory(settings);
    }

    public final void AddManager(String package_Keyword, String name, Manager model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getManagerTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddServieClient(String package_Keyword, String name, ServiceClient model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getServiceClientTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddServiceClientInterface(String name, ServiceClient model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(settings.getPackage(), name);
        Templates.getServiceClientInterfaceTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddServieClientBuilder(String package_Keyword, String name, ServiceClient model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getServiceClientBuilderTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddMethodGroup(String package_Keyword, String name, MethodGroupClient model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getMethodGroupTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddMethodGroupInterface(String name, MethodGroupClient model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(settings.getPackage(), name);
        Templates.getMethodGroupInterfaceTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddModel(String package_Keyword, String name, ClientModel model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getModelTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddException(String package_Keyword, String name, ClientException model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getExceptionTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddEnum(String package_Keyword, String name, EnumType model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getEnumTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddPage(String package_Keyword, String name, PageDetails model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getPageTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddClientResponse(String package_Keyword, String name, ClientResponse model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getResponseTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddXmlSequenceWrapper(String package_Keyword, String name, XmlSequenceWrapper model)
    {
        JavaFile javaFile = javaFileFactory.CreateSourceFile(package_Keyword, name);
        Templates.getXmlSequenceWrapperTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    public final void AddPackageInfo(String package_Keyword, String name, PackageInfo model)
    {
        JavaFile javaFile = javaFileFactory.CreateEmptySourceFile(package_Keyword, name);
        Templates.getPackageInfoTemplate().Write(model, javaFile);
        javaFiles.add(javaFile);
    }

    // TODO: POM?
//    public final void AddPom(PomTemplate pomTemplate)
//    {
//        StringBuilder pomContentsBuilder = new StringBuilder();
//        try (pomTemplate.TextWriter = new StringWriter(pomContentsBuilder))
//        {
//            pomTemplate.ExecuteAsync().GetAwaiter().GetResult();
//        }
//        javaFiles.add(new JavaFile("pom.xml", pomContentsBuilder.toString()));
//    }
}